### DISCUSSION TITLE
Expose `RequestContext.createContext()` for use in Elysia request decorator?

### DISCUSSION THREAD

**Discussion by crishoj** [mikro-orm/mikro-orm#4652] [Q&A] [answered]
[Elysia](https://elysiajs.com) doesn't follow the `app.use((req, res, next) => {})` pattern for middleware, so it's not immediately obvious how to properly associate each request with a fresh identity map fork.

[Decorating](https://elysiajs.com/concept/state-decorate.html) the request with a `RequestContext` might be an option:

```ts
const app = new Elysia()
    .decorate('ctx', RequestContext.createContext(orm.em) as RequestContext)
```

However, `.createContext()` is private in the type declaration.

Would it make sense to mark it public, or is there a better alternative that I'm unaware of?

**Comment by crishoj** [comment]
Perhaps it's as simple as `orm.em.fork()`?

```ts
const app = new Elysia()
    .decorate('em', orm.em.fork() as SqlEntityManager<SqliteDriver>)
```

**Comment by B4nan** [comment]
Exposing that method wouldn't help you much. Also, the `decorate` method accepts a callback, not a value. I believe you should use the global hooks, which is IMO just a fancy name for middleware. 

https://elysiajs.com/concept/life-cycle.html#global-hook

It seems to be basically the same as with fastify, so you can take a look here as well:

https://github.com/mikro-orm/guide#fastify-setup

```ts
const app = new Elysia();
app.on('request', () => RequestContext.create(orm.em, next));
```

This itself wont work, you will need to find out how it works with async hooks in elysia, their docs are very sparse. But I managed to find a mention that those hooks can be async, so there is definitely a way, worst case with a `new Promise()` to have the callback.

**Reply by crishoj** [reply]
@SaltyAom, would you like to chime in? 🙏

**Reply by vanlueckn** [reply]
How did you solve this? `RequestContext.create` expects `next`, but Elysia does not provide this.

**Reply by B4nan** [reply]
You could try this, but I am honestly not sure if its gonna help:

```ts
app.on('request', () => new Promise(resolve => RequestContext.create(orm.em, resolve));
```

Probably a more native way would be adding the ORM fork to the elysia context, so something like this:

```ts
app.on('request', ctx => ctx.em = orm.em.fork());
```

And then use it directly instead of relying on a global handle. If you would like to use some DI layer with repositories, then I would go with the `@CreateRequestContext` decorator.

**Reply by B4nan** [reply]
Or I guess you would go with your own AsyncLocalStorage and using `enterWith` as suggested in https://github.com/elysiajs/elysia/issues/109#issuecomment-1712765828

**Comment by B4nan** [comment, marked as answer]
I've added a new `RequestContext.enter()` method which doesn't have the `next` parameter and uses `ALS.enterWith()` instead of `run()`, it will be available in a few minutes via dev version, and later on with v6.3.5.

Here is a working example:

```ts
import { Elysia } from 'elysia';
import { EntitySchema, MikroORM, RequestContext } from '@mikro-orm/libsql';

export class User {

  id!: bigint;
  name: string;
  email: string;
  password: string;

  constructor(name: string, email: string, password: string) {
    this.name = name;
    this.email = email;
    this.password = password;
  }

}

export const schema = new EntitySchema({
  class: User,
  properties: {
    id: { type: 'bigint', primary: true },
    name: { type: 'string' },
    email: { type: 'string' },
    password: { type: 'string', lazy: true },
  },
});

const orm = MikroORM.initSync({
  dbName: ':memory',
  entities: [User],
});

const app = new Elysia();
app.on('request', () => RequestContext.enter(orm.em));
app.get('/', (req: Request, res: Response) => {
  console.log('hello from root', RequestContext.getEntityManager()?.id);
});
app.listen(3000);

console.log(
  `🦊 Elysia is running at ${app.server?.hostname}:${app.server?.port}`
);
```

**Reply by crishoj** [reply]
Thanks, I will try this once released!

**Reply by B4nan** [reply]
https://github.com/mikro-orm/elysia-bun-example-app

**Reply by crishoj** [reply]
Works perfectly — each request has its own EM:

<img width="350" alt="image" src="https://github.com/user-attachments/assets/c21a1737-3fd2-4f57-81bb-4a5bd0b087df">

Somewhat off-topic: What is the idiomatic way to manage EM/RequestContext for background tasks on the server?

**Reply by B4nan** [reply]
https://mikro-orm.io/docs/identity-map#createrequestcontext-decorator
